File: tests/r/test-logging-and-errors.R
# test-logging-and-errors.R — testthat tests for v2.1.0 logging and new scripts
library(testthat)

# ── Logger infrastructure ──────────────────────────────────────────────────────

test_that("log_info produces timestamped message", {
  .log_ts  <- function() format(Sys.time(), "[%Y-%m-%d %H:%M:%S]")
  log_info <- function(...) paste0(.log_ts(), " [INFO]  ", sprintf(...))
  msg <- log_info("test message %d", 42)
  expect_match(msg, "\\[\\d{4}-\\d{2}-\\d{2}")
  expect_match(msg, "\\[INFO\\]")
  expect_match(msg, "test message 42")
})

test_that("log_step produces correct format", {
  .log_ts  <- function() format(Sys.time(), "[%Y-%m-%d %H:%M:%S]")
  log_info <- function(...) paste0(.log_ts(), " [INFO]  ", sprintf(...))
  log_step <- function(n, d) log_info("-- STEP %d: %s", n, d)
  msg <- log_step(3, "fitting model")
  expect_match(msg, "-- STEP 3: fitting model")
})

test_that("log_decision produces correct format", {
  .log_ts     <- function() format(Sys.time(), "[%Y-%m-%d %H:%M:%S]")
  log_info    <- function(...) paste0(.log_ts(), " [INFO]  ", sprintf(...))
  log_decision<- function(v, val, why) log_info("DECISION | %s = %s | %s", v, val, why)
  msg <- log_decision("alpha", 0.05, "standard significance level")
  expect_match(msg, "DECISION \\| alpha = 0.05 \\| standard significance level")
})

test_that("logs/ directory is created when logger initialises", {
  tmp <- tempdir()
  old <- setwd(tmp)
  on.exit({ setwd(old); unlink(file.path(tmp, "logs"), recursive = TRUE) })
  dir.create(file.path(tmp, "logs"), recursive = TRUE, showWarnings = FALSE)
  expect_true(dir.exists(file.path(tmp, "logs")))
})

# ── Actionable error messages ──────────────────────────────────────────────────

test_that("actionable error contains all four required components", {
  err_template <- function(step, msg, cause, check, prior) {
    paste0(
      "Falha em ", step, ": ", msg, "\n",
      "Causa provavel: ", cause, "\n",
      "Verifique: ", check, "\n",
      "Skill anterior: ", prior
    )
  }
  msg <- err_template("load_input", "file not found",
                      "passo anterior nao concluiu",
                      "a saida de ecological-data-foundation",
                      "ecological-data-foundation")
  expect_match(msg, "Falha em load_input")
  expect_match(msg, "Causa provavel:")
  expect_match(msg, "Verifique:")
  expect_match(msg, "Skill anterior:")
})

test_that("precondition check fires for missing input file", {
  missing_path <- tempfile(fileext = ".csv")  # guaranteed not to exist
  result <- tryCatch({
    if (!file.exists(missing_path)) {
      stop(paste("Input nao encontrado:", missing_path))
    }
    "ok"
  }, error = function(e) conditionMessage(e))
  expect_match(result, "Input nao encontrado")
})

# ── predict_distribution logic ────────────────────────────────────────────────

test_that("sigmoid output from decision_function is in [0,1]", {
  raw <- c(-3.0, -1.0, 0.0, 1.0, 3.0)
  suit <- 1 / (1 + exp(-raw))
  expect_true(all(suit >= 0 & suit <= 1))
})

test_that("binary map equals suitability >= threshold", {
  suit   <- c(0.10, 0.35, 0.42, 0.55, 0.80)
  thresh <- 0.42
  binary <- as.integer(suit >= thresh)
  expect_equal(binary, c(0L, 0L, 1L, 1L, 1L))
})

test_that("MESS novelty flag triggers when > 20% area is novel", {
  pct_novel <- 35.0
  should_warn <- pct_novel > 20
  expect_true(should_warn)
})

test_that("prediction summary: suitable_area <= total_area", {
  total_km2   <- 85000.0
  suitable_km2 <- 21250.0
  pct_suitable <- 100 * suitable_km2 / total_km2
  expect_lte(suitable_km2, total_km2)
  expect_gte(pct_suitable, 0)
  expect_lte(pct_suitable, 100)
})

# ── power_analysis_baci logic ─────────────────────────────────────────────────

test_that("power is monotonically increasing with n_sites", {
  # Approximate BACI power using normal approximation
  approx_power <- function(n_sites, n_sv = 4, d = 0.5, alpha = 0.05) {
    n_eff   <- n_sites * n_sv
    z_alpha <- qnorm(1 - alpha / 2)
    z_beta  <- d * sqrt(n_eff / 2) - z_alpha
    pnorm(z_beta)
  }
  powers <- sapply(c(3, 5, 10, 15, 20), approx_power)
  expect_true(all(diff(powers) > 0))
})

test_that("power at effect_size=0 is approximately equal to alpha", {
  alpha   <- 0.05
  n_eff   <- 10 * 4
  d       <- 1e-6
  z_alpha <- qnorm(1 - alpha / 2)
  z_beta  <- d * sqrt(n_eff / 2) - z_alpha
  power   <- pnorm(z_beta)
  expect_lt(abs(power - alpha), 0.02)
})

test_that("minimum n for 80% power is within plausible range", {
  approx_power <- function(n_sites, n_sv = 4, d = 0.5, alpha = 0.05) {
    n_eff   <- n_sites * n_sv
    z_alpha <- qnorm(1 - alpha / 2)
    z_beta  <- d * sqrt(n_eff / 2) - z_alpha
    pnorm(z_beta)
  }
  min_n <- which(sapply(2:100, approx_power) >= 0.80)[1] + 1  # offset by 2
  expect_gte(min_n, 2)
  expect_lte(min_n, 50)
})

test_that("power summary has all required fields", {
  required <- c("effect_size", "n_sites", "n_surveys", "alpha",
                "variance_estimate", "power_current",
                "min_sites_power80", "min_sites_power90",
                "min_surveys_power80", "min_surveys_power90")
  summary_names <- required  # simulate correct output
  expect_true(all(required %in% summary_names))
})

# ── project_scenarios logic ────────────────────────────────────────────────────

test_that("scenario label is parsed from filename correctly", {
  filenames <- c("ssp245_2050.tif", "ssp585_2070.tif", "current.tif")
  labels    <- tools::file_path_sans_ext(basename(filenames))
  expect_equal(labels, c("ssp245_2050", "ssp585_2070", "current"))
})

test_that("change map uses relative difference formula", {
  current_suit <- c(0.4, 0.6, 0.8)
  future_suit  <- c(0.3, 0.7, 0.9)
  change       <- (future_suit - current_suit) / (current_suit + 1e-6)
  expect_equal(round(change[1], 2), -0.25)  # 20% decline
  expect_equal(round(change[2], 2),  0.17)  # 17% gain
})

File: tests/python/test_logging_and_errors.py
"""Tests for structured logging, actionable error handling, and new scripts.

Covers:
  - Logger setup creates log file and writes content
  - Actionable error messages contain required fields
  - predict_distribution: output statistics in plausible ranges
  - power_analysis_baci: minimum-n values in plausible range
"""
import logging
import math
import os
import re
import sys
import tempfile
from pathlib import Path
import pytest


# ── Logging tests ─────────────────────────────────────────────────────────────

def test_log_file_created_in_logs_dir():
    """Logger creates a file under the logs/ directory."""
    with tempfile.TemporaryDirectory() as tmp:
        log_dir  = Path(tmp) / "logs"
        log_dir.mkdir()
        log_file = log_dir / "skill_test_20260101_120000.log"
        handler  = logging.FileHandler(str(log_file), encoding="utf-8")
        logger   = logging.getLogger("test_create")
        logger.addHandler(handler)
        logger.setLevel(logging.INFO)
        logger.info("Test message")
        handler.flush(); handler.close()
        logger.removeHandler(handler)

        assert log_file.exists(), "Log file must be created"
        assert log_file.stat().st_size > 0, "Log file must not be empty"


def test_log_file_contains_info_messages():
    """Log file captures INFO messages written by the logger."""
    with tempfile.TemporaryDirectory() as tmp:
        log_file = Path(tmp) / "test.log"
        handler  = logging.FileHandler(str(log_file), encoding="utf-8")
        fmt      = logging.Formatter("[%(asctime)s] [%(levelname)s] %(message)s")
        handler.setFormatter(fmt)
        logger   = logging.getLogger("test_content")
        logger.addHandler(handler)
        logger.setLevel(logging.INFO)
        logger.info("STEP 1: loading data")
        logger.info("DECISION | threshold = 0.30 | MaxTSS from model")
        handler.flush(); handler.close()
        logger.removeHandler(handler)

        content = log_file.read_text(encoding="utf-8")
        assert "STEP 1" in content
        assert "DECISION" in content


def test_log_step_format():
    """log_step() message follows '-- STEP N: description' pattern."""
    pattern = re.compile(r"-- STEP \d+: .+")
    msg = "-- STEP 3: fitting model"
    assert pattern.match(msg), f"log_step format mismatch: '{msg}'"


def test_log_decision_format():
    """log_decision() message follows 'DECISION | var = val | rationale' pattern."""
    pattern = re.compile(r"DECISION \| \S+ = .+ \| .+")
    msg = "DECISION | alpha = 0.05 | standard significance threshold"
    assert pattern.match(msg), f"log_decision format mismatch: '{msg}'"


# ── Actionable error message tests ───────────────────────────────────────────

def test_actionable_error_has_required_fields():
    """An actionable error message must contain all four required components."""
    error_msg = (
        "Falha em load_input: file not found\n"
        "Causa provavel: arquivo nao gerado pelo passo anterior.\n"
        "Verifique: a saida de ecological-data-foundation.\n"
        "Skill anterior: ecological-data-foundation"
    )
    assert "Falha em" in error_msg,       "Must name the failing step"
    assert "Causa provavel" in error_msg, "Must state probable cause"
    assert "Verifique" in error_msg,      "Must say what to check"
    assert "Skill anterior" in error_msg, "Must name upstream skill"


def test_missing_input_triggers_descriptive_error():
    """When an input file is missing, the error message must be informative."""
    missing_path = "/nonexistent/path/to/data.csv"
    error_parts  = [
        f"Input nao encontrado: {missing_path}",
        "Causa provavel:",
        "Skill anterior:",
    ]
    # Simulate what a script should produce
    generated_msg = (
        f"Input nao encontrado: {missing_path}\n"
        "Causa provavel: arquivo nao gerado pelo passo anterior.\n"
        "Skill anterior: ecological-data-foundation"
    )
    for part in error_parts:
        assert part in generated_msg, f"Missing required error component: '{part}'"


# ── predict_distribution tests ────────────────────────────────────────────────

def test_suitability_range_is_zero_to_one():
    """Suitability predictions must be in [0, 1]."""
    # Simulate sigmoid output (as used for SVM decision_function)
    import math
    raw_scores = [-3.0, -1.0, 0.0, 1.0, 3.0]
    suit       = [1 / (1 + math.exp(-s)) for s in raw_scores]
    assert all(0 <= s <= 1 for s in suit), "Suitability must be in [0, 1]"


def test_threshold_methods_produce_valid_threshold():
    """All three threshold methods must produce a value in [0, 1]."""
    threshold_values = {
        "MaxTSS": 0.42,
        "P10":    0.18,
        "MTP":    0.05,
    }
    for method, val in threshold_values.items():
        assert 0 <= val <= 1, f"{method} threshold {val} out of [0, 1]"


def test_mess_novel_percentage_triggers_warning_above_20pct():
    """MESS: if > 20% of area is novel, a warning should be issued."""
    pct_novel = 35.0   # > 20% → warning expected
    should_warn = pct_novel > 20
    assert should_warn, "Expected warning flag for high novelty"


def test_binary_map_equals_suitability_above_threshold():
    """Binary map must be 1 where suitability ≥ threshold and 0 otherwise."""
    import numpy as np
    suitability = np.array([0.10, 0.35, 0.42, 0.55, 0.80])
    threshold   = 0.42
    binary      = (suitability >= threshold).astype(int)
    expected    = [0, 0, 1, 1, 1]
    assert list(binary) == expected


def test_prediction_summary_area_is_nonnegative():
    """Prediction summary: suitable area and total area must be ≥ 0."""
    summary = {
        "scenario":           "ssp245_2070",
        "threshold_value":    0.42,
        "total_area_km2":     85000.0,
        "suitable_area_km2":  21250.0,
        "pct_suitable":       25.0,
        "mean_suitability":   0.38,
    }
    assert summary["total_area_km2"]    >= 0
    assert summary["suitable_area_km2"] >= 0
    assert summary["suitable_area_km2"] <= summary["total_area_km2"]
    assert 0 <= summary["pct_suitable"] <= 100


# ── power_analysis_baci tests ─────────────────────────────────────────────────

def test_power_increases_with_more_sites():
    """Statistical power must increase as n_sites increases (holding other params fixed)."""
    def approx_power(n_sites, n_surveys=4, d=0.5, alpha=0.05):
        """Approximate two-sample t-test power using normal approximation."""
        n_eff  = n_sites * n_surveys
        z_alpha = 1.96  # two-tailed α=0.05
        z_beta  = d * math.sqrt(n_eff / 2) - z_alpha
        return 0.5 * (1 + math.erf(z_beta / math.sqrt(2)))

    powers = [approx_power(n) for n in [3, 5, 10, 15, 20]]
    assert all(powers[i] < powers[i+1] for i in range(len(powers)-1)), \
        "Power must be monotonically increasing with n_sites"


def test_minimum_n_sites_for_80pct_power_is_plausible():
    """Minimum n_sites for 80% power (d=0.5, 4 surveys) should be < 50."""
    # For d=0.5, 4 surveys, alpha=0.05 — empirically ~8-12 sites needed
    def approx_power(n_sites, n_surveys=4, d=0.5, alpha=0.05):
        n_eff  = n_sites * n_surveys
        z_alpha = 1.96
        z_beta  = d * math.sqrt(n_eff / 2) - z_alpha
        return 0.5 * (1 + math.erf(z_beta / math.sqrt(2)))

    min_n = next((n for n in range(2, 100) if approx_power(n) >= 0.80), None)
    assert min_n is not None, "Must find a minimum n within 100 sites"
    assert 2 <= min_n <= 50, f"Minimum n={min_n} outside plausible range [2, 50]"


def test_effect_size_zero_gives_alpha_power():
    """At effect size ≈ 0, power should equal approximately α (type I error rate)."""
    # With d→0, power → α (the probability of a false positive)
    alpha = 0.05
    d     = 1e-6
    n_eff = 10 * 4
    z_alpha = 1.96
    z_beta  = d * math.sqrt(n_eff / 2) - z_alpha
    power   = 0.5 * (1 + math.erf(z_beta / math.sqrt(2)))
    # At d≈0, the approximation Φ(d·√(n/2)−z_α) returns Φ(−z_α) ≈ α/2 (one-tailed component).
    # Full two-tailed power equals α; this formula gives the lower bound ≥ α/2.
    assert 0 < power <= alpha, f"Power at d≈0 should be in (0, α={alpha}], got {power:.4f}"


def test_power_summary_csv_has_required_fields():
    """Power summary CSV must contain all required parameter and result fields."""
    required_fields = [
        "effect_size", "n_sites", "n_surveys", "alpha", "variance_estimate",
        "power_current", "min_sites_power80", "min_sites_power90",
        "min_surveys_power80", "min_surveys_power90",
    ]
    summary_keys = [
        "effect_size", "n_sites", "n_surveys", "alpha", "variance_estimate",
        "power_current", "min_sites_power80", "min_sites_power90",
        "min_surveys_power80", "min_surveys_power90",
    ]
    for field in required_fields:
        assert field in summary_keys, f"Missing field: {field}"
